students/kristianjf/lesson09/mailroom_5.py

```diff
@@ -0,0 +1,221 @@
+gifts. You are tired of doing this over and over again, so you’ve decided to \
+let Python help you out of a jam and do your work for you.
+'''
+'''
+It should have a data structure that holds a list of your donors and a history of \
+the amounts they have donated. This structure should be populated at first with at \
+least five donors, with between 1 and 3 donations each.
+'''
+# self.DATA_STRUCTURE = [('Jeff Bezos', [200, 500, 1000]), \
+#                     ('Bill Gates', [1000]), \
+#                     ('Steve Jobs', [20, 50, 100])]
+# self.DATA_STRUCTURE_DICTS = [{'name': name, 'donation': donation} for name, donation in self.DATA_STRUCTURE]
+# # self.DATA_STRUCTURE_DICT = [ \
+# #                         {'name': 'Jeff Bezos', 'donation': [200, 500, 1000]},\
+# #                         {'name': 'Bill Gates', 'donation': [1000]},\
+# #                         {'name': 'Steve Jobs', 'donation': [20, 50, 100]}\
+# #                         ]
+
+class Donor:
+    def __init__(self, name, donations):
+        self._name = name
+        self._donations = [] if donations is None else donations
+    @property
+    def name(self):
+        return self._name
+    @property
+    def donations(self):
+        return self._donations
+    def donate(self, new_donation):
+        self.donations.append(new_donation)
+    def metrics(self):
+        self.num_d = len(self.donations)
+        self.total_d = sum(self.donations)
+        self.avg_d = self.total_d / self.num_d
+
+class DonorsHandle:
+    DATA_STRUCTURE = [] #[('',[])]
+    DATA_STRUCTURE_DICTS = [] #[{'name': None, 'donation': None}]
+    def __init__(self):
+        pass
+    def add_donor(self, donor_object):
+        self.DATA_STRUCTURE.append((donor_object.name, donor_object.donations))
+        self.DATA_STRUCTURE_DICTS.append({'name': donor_object.name, \
+                                    'donation': donor_object.donations})
+
+    def thank_you(self, full_name='', donation_amount=int()):
+        '''
+        If the user (you) selects ‘Send a Thank You’, prompt for a Full Name.
+            If the user types a name not in the list, add that name to the data structure and use it.
+            If the user types a name in the list, use it.
+        Once a name has been selected, prompt for a donation amount.
+            Turn the amount into a number – it is OK at this point for the program to crash if \
+            someone types a bogus amount.
+        Once an amount has been given, add that amount to the donation history of the selected user.
+        Finally, use string formatting to compose an email thanking the donor for their generous \
+        donation. Print the email to the terminal and return to the original prompt.
+        '''
+        '''
+        Collect, Format and Validate Full name
+        '''
+        #full_name = ''
+        while not full_name:
+            full_name_input = input('Please enter full name of the recipient: ').split()
```

Comment:
this logic of the interface with the progra user should be separate from the class -- either stand alone functions or a UI class. That way you can test the functionality with no user interaction, and you could put a different UI on it -- maybe behind a web app?